src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs

```diff
@@ -97,21 +98,35 @@ public Task Invoke(HttpContext context)
                     case StaticFileContext.PreconditionState.ShouldProcess:
                         if (fileContext.IsHeadMethod)
                         {
-                            return fileContext.SendStatusAsync(Constants.Status200Ok);
+                            await fileContext.SendStatusAsync(Constants.Status200Ok);
+                            return;
                         }
                         if (fileContext.IsRangeRequest)
                         {
-                            return fileContext.SendRangeAsync();
+                            await fileContext.SendRangeAsync();
+                            return;
+                        }
+                        try
+                        {
+                            await fileContext.SendAsync();
+                            _logger.LogFileServed(fileContext.SubPath, fileContext.PhysicalPath);
+                            return;
+                        }
+                        catch (FileNotFoundException)
+                        {
+                            _logger.LogFileNotFound(fileContext.SubPath);
+                            await fileContext.SendStatusAsync(StatusCodes.Status404NotFound);
```

Comment:
Returning an explicit 404 is a strange choice. There's no other condition where this middleware returns 404. Calling next would be better.